### PR TITLE
Send A Package Entity

### PR DESCRIPTION
**Pull request by abdul-ahad12**
## Description

Created a new entity for the SendAPackage service, removed the previous Order module, and generated a fresh migration script to reflect the updated changes.

### CODE REVIEW

**Review by amitravikumar — APPROVED** [review]

File: apps/backend/backend-main/src/entity/Report.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+})
+@Index('IDX_report_orderId', ['sendPackageOrderId'], {
+  where: '"deletedAt" IS NULL',
+})
+@Entity()
+export class Report extends BaseEntity {
+  @Column({ type: 'varchar', nullable: false })
+  reason: string;
+
+  @Column({ type: 'text', nullable: true })
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/Report.ts:17]: type: 'text' -> type: 'varchar', this is to maintain consistent behavior across the entity files.

File: apps/backend/backend-main/src/entity/Report.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  details: string;
+
+  @ManyToOne(() => User, {
+    cascade: true,
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/Report.ts:21]: redundant 🪓  as you have already used onDelete: 'CASCADE'

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  @Column({
+    type: 'varchar',
+    default: OrderTypeEnum.DELIVERY,
+    nullable: false,
+  })
+  type: OrderTypeEnum;
+
+  @OneToOne(() => PickupLocation, {
+    cascade: true,
+    onDelete: 'CASCADE',
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:58]: , deferrable: 'INITIALLY IMMEDIATE',

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  @JoinColumn()
+  pickupLocation: PickupLocation;
+
+  @RelationId((sendPackageOrder: SendPackageOrder) => sendPackageOrder.pickupLocation)
+  @Column({ type: 'integer' })
+  pickupLocationId: number;
+
+  @OneToOne(() => DropLocation, {
+    cascade: true,
+    onDelete: 'CASCADE',
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:69]: , deferrable: 'INITIALLY IMMEDIATE'

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  })
+  @JoinColumn()
+  dropLocation: DropLocation;
+
+  @RelationId((sendPackageOrder: SendPackageOrder) => sendPackageOrder.dropLocation)
+  @Column({ type: 'integer' })
+  dropLocationId: number;
+
+  @Column({ nullable: false })
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:78]: type?

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  estimatedDistance: number;
+
+  @Column({ nullable: false })
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:81]: type?

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  estimatedTime: number;
+
+  @ManyToOne(() => User, {
+    cascade: true,
+    onDelete: 'CASCADE',
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:86]: ,deferrable: 'INITIALLY IMMEDIATE',

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  })
+  @JoinColumn()
+  pickupLocation: PickupLocation;
+
+  @RelationId((sendPackageOrder: SendPackageOrder) => sendPackageOrder.pickupLocation)
+  @Column({ type: 'integer' })
+  pickupLocationId: number;
+
+  @OneToOne(() => DropLocation, {
+    cascade: true,
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:68]: redundant 🪓

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+  @Column({
+    type: 'varchar',
+    default: OrderTypeEnum.DELIVERY,
+    nullable: false,
+  })
+  type: OrderTypeEnum;
+
+  @OneToOne(() => PickupLocation, {
+    cascade: true,
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:57]: redundant 🪓

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  dropLocationId: number;
+
+  @Column({ nullable: false })
+  estimatedDistance: number;
+
+  @Column({ nullable: false })
+  estimatedTime: number;
+
+  @ManyToOne(() => User, {
+    cascade: true,
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:85]: redundant 🪓

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  cancellationReason: string;
+
+  @Column({
+    type: 'varchar',
+    nullable: true,
+  })
+  canceledBy: UserRoleEnum;
+
+  @ManyToOne(() => Agent, {
+    cascade: true,
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:124]: redundat 🪓 

add deferrable: 'INITIALLY IMMEDIATE',

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    onDelete: 'CASCADE',
+    nullable: true,
+  })
+  assignedAgent: Agent;
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:128]: Pointless to add two many-to-one relations from the same entity `Agent` 🤷‍♂️  So 🪓  one relation

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+  @Column({
+    type: 'varchar',
+    default: PaymentStatusEnum.NOT_PAID,
+    nullable: false,
+  })
+  paymentStatus: PaymentStatusEnum;
+
+  @OneToOne(() => Report, (report) => report.sendPackageOrder, {
+    cascade: true,
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:154]: redundant 🪓 , typeorm will handle this on its own.

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+  @Column({ type: 'integer', nullable: true })
+  assignedAgentId: number;
+
+  @Column({ type: 'text', nullable: true })
+  completionPhoto: string;
+
+  @Column({ type: 'timestamp', nullable: true })
+  acceptedAt: Date;
+
+  @Column({ type: 'timestamp', nullable: true })
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:140]: can this be `null`?

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+  @Column({ type: 'float', nullable: true })
+  actualTime: number;
+
+  @Column({ type: 'text', nullable: true })
+  cancellationReason: string;
+
+  @Column({
+    type: 'varchar',
+    nullable: true,
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:119]: can this be `null` ?

Do we want to use any default value from the enum if any are applicable?

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+  @Column({ type: 'integer', nullable: true })
+  agentId: number;
+
+  @Column({ type: 'float', nullable: true })
+  price: number;
+
+  @Column({ type: 'float', nullable: true })
+  actualDistance: number;
+
+  @Column({ type: 'float', nullable: true })
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:111]: can these critical values be `null` ?

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+    onDelete: 'CASCADE',
+  })
+  customer: User;
+
+  @RelationId((sendPackageOrder: SendPackageOrder) => sendPackageOrder.customer)
+  @Column({ type: 'integer' })
+  customerId: number;
+
+  @ManyToOne(() => Agent, {
+    cascade: true,
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:95]: redundant 🪓

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+    type: 'varchar',
+    default: PaymentStatusEnum.NOT_PAID,
+    nullable: false,
+  })
+  paymentStatus: PaymentStatusEnum;
+
+  @OneToOne(() => Report, (report) => report.sendPackageOrder, {
+    cascade: true,
+    nullable: true,
+    onDelete: 'SET NULL',
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:156]: Good use of this 👍

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+
+  @Column({
+    type: 'varchar',
+    nullable: true,
+  })
+  canceledBy: UserRoleEnum;
+
+  @ManyToOne(() => Agent, {
+    cascade: true,
+    onDelete: 'CASCADE',
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:125]: onDelete: 'SET NULL',

Use the above as an `Agent` will come and go anytime from the business, so whenever an `Agent` quits the job, this record should not be deleted.

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts (hunk 7 above)
Comment (theshdb) [reply apps/backend/backend-main/src/entity/SendPackageOrder.ts:86]: same for customer as well: https://github.com/novareach-solutions/movemate/pull/65/files#r1881388048

Comment (theshdb) [reply apps/backend/backend-main/src/entity/SendPackageOrder.ts:86]: also add `nullable`: true

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts (hunk 4 above)
Comment (theshdb) [reply apps/backend/backend-main/src/entity/SendPackageOrder.ts:69]: use onDelete: 'SET NULL'

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts (hunk 3 above)
Comment (theshdb) [reply apps/backend/backend-main/src/entity/SendPackageOrder.ts:58]: Use onDelete: 'SET NULL'

Comment (theshdb) [reply apps/backend/backend-main/src/entity/SendPackageOrder.ts:58]: Add nullable: true

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts (hunk 4 above)
Comment (theshdb) [reply apps/backend/backend-main/src/entity/SendPackageOrder.ts:69]: Add nullable: true

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+  status: OrderStatusEnum;
+
+  @Column({
+    type: 'varchar',
+    default: OrderTypeEnum.DELIVERY,
+    nullable: false,
+  })
+  type: OrderTypeEnum;
+
+  @OneToOne(() => PickupLocation, {
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:56]: This should be Many to One; Many `SendPackageOrder` can have one `PickUpLocation`

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+    onDelete: 'CASCADE',
+  })
+  @JoinColumn()
+  pickupLocation: PickupLocation;
+
+  @RelationId((sendPackageOrder: SendPackageOrder) => sendPackageOrder.pickupLocation)
+  @Column({ type: 'integer' })
+  pickupLocationId: number;
+
+  @OneToOne(() => DropLocation, {
```
Comment (theshdb) [inline apps/backend/backend-main/src/entity/SendPackageOrder.ts:67]: This should be Many to One; Many `SendPackageOrder` can have one `DropLocation`

Basically, ordering from the same person who has already set an address will not do it again.

File: apps/backend/backend-main/src/entity/SendPackageOrder.ts (hunk 14 above)
Comment (abdul-ahad12) [reply apps/backend/backend-main/src/entity/SendPackageOrder.ts:140]: Yes, while creating the order it is going to be null!

**Pull request merged**